### ISSUE TITLE
MIPS: Instruction Disassembly/Lifting Completion

### ISSUE THREAD

**Issue by plafosse** [Vector35/binaryninja-docs#212]
The following is the list of instructions which we currently disassemble and lift (Fully or Partially). If you have any instructions which differ from this table, there is likely a bug or a documentation failure, please let us know (also if you could provide the opcodes that would be great).

| Mnem | Disasm | Lifting |
|------|--------|---------|
| abss | Full | Full |
| add | Full | Full |
| addu | Full | Full |
| and | Full | Full |
| ceilwd | Full | Full |
| ceilws | Full | Full |
| ceqd | Full | Full |
| ceqs | Full | Full |
| cfd | Full | Full |
| cfs | Full | Full |
| clo | Full | Full |
| clz | Full | Full |
| cngld | Full | Full |
| cngled | Full | Full |
| cngles | Full | Full |
| cngls | Full | Full |
| coled | Full | Full |
| coles | Full | Full |
| coltd | Full | Full |
| colts | Full | Full |
| cseqd | Full | Full |
| cseqs | Full | Full |
| csfd | Full | Full |
| csfs | Full | Full |
| cueqd | Full | Full |
| cueqs | Full | Full |
| culed | Full | Full |
| cules | Full | Full |
| cultd | Full | Full |
| cults | Full | Full |
| cund | Full | Full |
| cuns | Full | Full |
| cvtds | Full | Full |
| cvtdw | Full | Full |
| cvtsd | Full | Full |
| cvtsw | Full | Full |
| cvtwd | Full | Full |
| floorwd | Full | Full |
| floorws | Full | Full |
| ins | Full | Full |
| ldc2 | Full | Full |
| lwc2 | Full | Full |
| lwxc1 | Full | Full |
| madd | Full | Full |
| maddu | Full | Full |
| movd | Full | Full |
| move | Full | Full |
| movs | Full | Full |
| neg | Full | Full |
| negs | Full | Full |
| negu | Full | Full |
| nor | Full | Full |
| not | Full | Full |
| or | Full | Full |
| roundwd | Full | Full |
| roundws | Full | Full |
| slt | Full | Full |
| sltu | Full | Full |
| sqrts | Full | Full |
| sub | Full | Full |
| subu | Full | Full |
| suxc1 | Full | Full |
| swxc1 | Full | Full |
| truncwd | Full | Full |
| truncws | Full | Full |
| wsbh | Full | Full |
| xor | Full | Full |
| absd | Full | None |
| addd | Full | None |
| addi | Full | None |
| addiu | Full | None |
| adds | Full | None |
| andi | Full | None |
| cfc1 | Full | None |
| cled | Full | None |
| cles | Full | None |
| cltd | Full | None |
| clts | Full | None |
| cnged | Full | None |
| cnges | Full | None |
| cngtd | Full | None |
| cngts | Full | None |
| ctc1 | Full | None |
| cvtws | Full | None |
| lb | Full | None |
| lbu | Full | None |
| lh | Full | None |
| lhu | Full | None |
| ll | Full | None |
| lui | Full | None |
| luxc1 | Full | None |
| lw | Full | None |
| mfc0 | Full | None |
| mfc1 | Full | None |
| mfc2 | Full | None |
| mfhc1 | Full | None |
| mfhi | Full | None |
| mflo | Full | None |
| movf | Full | None |
| movfd | Full | None |
| movfs | Full | None |
| movt | Full | None |
| msub | Full | None |
| msubu | Full | None |
| mtc0 | Full | None |
| mtc1 | Full | None |
| mtc2 | Full | None |
| mthc1 | Full | None |
| mthi | Full | None |
| mtlo | Full | None |
| muld | Full | None |
| muls | Full | None |
| mult | Full | None |
| multu | Full | None |
| negd | Full | None |
| ori | Full | None |
| rotr | Full | None |
| rotrv | Full | None |
| sb | Full | None |
| sc | Full | None |
| sdc2 | Full | None |
| sh | Full | None |
| sll | Full | None |
| sllv | Full | None |
| slti | Full | None |
| sltiu | Full | None |
| sqrtd | Full | None |
| sra | Full | None |
| srav | Full | None |
| srl | Full | None |
| srlv | Full | None |
| subd | Full | None |
| subs | Full | None |
| sw | Full | None |
| swc1 | Full | None |
| swc2 | Full | None |
| swl | Full | None |
| xori | Full | None |

**Comment by aschmitz** [comment]
#261 and #263 may just belong here instead.

**Comment by aschmitz** [comment]
Bug: `sltu` doesn't properly set value to 0 in LLIR.

In disassembling `00 62 18 2b`, binja correctly gets `sltu $v1, $v1, $v0`, but the LLIR tree looks instead like:

~~~
if ($v1 u< $v0) {
  $v1 = 1
} else {
  $v1 = $v0
}
~~~

The real interpretation in the "else" case should be `$v1 = 0`.

**Comment by Arcnor** [comment]
This list is missing `lwl` (Load Word Left) and `lwr` (Load Word Right) as I mentioned on #1140

**Comment by Arcnor** [comment]
It's also missing (at least) `cop2` from the coprocessor, as mentioned on #1131

**Comment by psifertex** [comment]
Thanks for aggregating these -- makes it easier to track when we're working on them. If there's one in particular that is slowing you down versus the others please let us know, or you can subclass our architecture and implement it yourself even. See, eg: https://github.com/Vector35/binaryninja-api/tree/dev/examples/x86_extension

**Comment by Arcnor** [comment]
Yep, just found that you had this generic issue. In my case the missing lifting is breaking the whole lifting for the function, so that one is the one that matters me the most.

I just found out that you can do it yourself as you mention, so I might try that soon.

**Comment by zznop** [comment]
`daddu` and `ld` are not being lifted (current stable 1.3)

**Comment by psifertex** [comment]
Thanks -- updated the original list.

**Comment by zznop** [comment]
`dsll32` doesn't even disassemble.

**Comment by nshp** [comment]
Seems like the MIPS coprocessor registers aren't classified as "system registers", so e.g. HLIL optimizes a function full of side-effectful `mtc0` and `mfc0` instructions down to nothing.

Also, maybe `ehb` should be lifted to an intrinsic?

**Comment by nshp** [comment]
The CACHE instruction OP is not decoded correctly.
Examples:
- `bc 85 00 00` is `cache 5, 0($a0)`, but binja shows `cache load, ($a0)`
- `bc 88 00 00` is `cache 8, 0($a0)`, but binja still shows `cache load, ($a0)`
- `bc 89 00 00` is `cache 9, 0($a0)`, but binja still shows `cache load, ($a0)`

The cache op is 5 bits, 20:16, split into two subfields:
- 17:16: Which cache
- 20:18: Operation

The above cases were (if my numbers are right):
- 5: data cache load tag
- 8: instruction cache store tag
- 9: data cache store tag
